# T05 - Motor Trend Car Road Tests

### 1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

- Calcula el R2 e interpreta los signos de los betas.
- Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.
- Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

In [1]:
#EDA

import pandas as pd

df = pd.read_excel("Motor Trend Car Road Tests.xlsx")

print(df.head())
print(df.describe())
print(df.info())

               model   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2  
             mpg        cyl        disp          hp       drat         wt  \
count  32.000000  32.000000   32.000000   32.000000  32.000000  32.000000   
mean   20.090625   6.187500  230.721875  146.687500   3.596563   3.217250   
std     6.026948   1.785922  123.938694   68.562868   0.534679   0.978457   
min    10.400000   4.000000   71.100000   52.000000   2.760000   1.513000   
25%    15.425000   4.000000  120.825000   96.500000   3.080000   2.5

In [3]:
#Definir "x", "y"

from sklearn.model_selection import train_test_split

X = df.drop(columns = ["model", "mpg"])
y = df['mpg'] 

In [5]:
# Escalamiento de variables x

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X)

X_scaled = scaler.transform(X)

In [7]:
# Regresión

from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm

# Agregar constante para el término independiente
X_scaled_const = sm.add_constant(X_scaled)

# Modelo 1
modelo_1 = sm.OLS(y, X_scaled_const).fit()

print(modelo_1.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.869
Model:                            OLS   Adj. R-squared:                  0.807
Method:                 Least Squares   F-statistic:                     13.93
Date:                Fri, 12 Sep 2025   Prob (F-statistic):           3.79e-07
Time:                        21:21:38   Log-Likelihood:                -69.855
No. Observations:                  32   AIC:                             161.7
Df Residuals:                      21   BIC:                             177.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         20.0906      0.468     42.884      0.0

###  Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

In [9]:
#Definir "x", "y"/ Separar en train y test

from sklearn.model_selection import train_test_split

X = df.drop(columns = ["model", "mpg"])
y = df['mpg']  

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4, random_state=137)

In [11]:
# Escalamiento de variables x

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X_train)

X_scaled_train = scaler.transform(X_train)

X_scaled_test = scaler.transform(X_test)

In [13]:
# Regresión

from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm

# Agregar constante para el término independiente
X_scaled_train_const = sm.add_constant(X_scaled_train)
X_scaled_test_const = sm.add_constant(X_scaled_test)

# Modelo train
modelo_train = sm.OLS(y_train, X_scaled_train_const).fit()

print(modelo_train.summary())

# Evaluación del modelo entrenado con TRAIN en TEST

y_pred_test = modelo_train.predict(X_scaled_test_const)

r2 = r2_score(y_test, y_pred_test)


print("")
print("")
print("")
print("")
print(f"R² de test: {r2:.4f}")

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.891
Method:                 Least Squares   F-statistic:                     9.981
Date:                Fri, 12 Sep 2025   Prob (F-statistic):              0.242
Time:                        21:22:29   Log-Likelihood:                -9.9811
No. Observations:                  12   AIC:                             41.96
Df Residuals:                       1   BIC:                             47.30
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333      0.556     31.361      0.0

C:\Users\juanm\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)


### Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

In [15]:
from sklearn.linear_model import Ridge

# Entrenar el modelo Ridge 
modelo_ridge = Ridge(alpha=16)
modelo_ridge.fit(X_scaled_train, y_train)

# Predicciones
y_train_pred_ridge = modelo_ridge.predict(X_scaled_train)
y_test_pred_ridge = modelo_ridge.predict(X_scaled_test)

# R2
r2_train_ridge = r2_score(y_train, y_train_pred_ridge)
r2_test_ridge = r2_score(y_test, y_test_pred_ridge)

# Resultados Ridge
print("Coeficientes Ridge:", modelo_ridge.coef_)
print("Intercepto Ridge:", modelo_ridge.intercept_)
print(f"R² en entrenamiento: {r2_train_ridge:.4f}")
print(f"R² en prueba:        {r2_test_ridge:.4f}")


Coeficientes Ridge: [-0.90402285 -0.75512585 -0.47375041  0.6741384  -0.71223975  0.46078294
  0.44021965  0.62513287  0.13478138 -0.4323926 ]
Intercepto Ridge: 17.433333333333334
R² en entrenamiento: 0.8355
R² en prueba:        0.7648


# 1.2 Repite el ejercicio anterior usando 'qsec' como salida.

### Modelo sin train / test

In [17]:
#EDA

import pandas as pd

df2 = pd.read_excel("Motor Trend Car Road Tests.xlsx")

# df2.head()
# df2.describe()
# df2.info()

#Definir "x", "y"/ Separar en train y test

from sklearn.model_selection import train_test_split

X = df2.drop(columns = ["model", "qsec"])
y = df2['qsec'] 

# Escalamiento de variables x

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X)

X_scaled = scaler.transform(X)

# Regresión

from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm

# Agregar constante para el término independiente
X_scaled_const = sm.add_constant(X_scaled)

# Modelo 1.2
modelo_12 = sm.OLS(y, X_scaled_const).fit()

print(modelo_12.summary())

                            OLS Regression Results                            
Dep. Variable:                   qsec   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.815
Method:                 Least Squares   F-statistic:                     14.66
Date:                Fri, 12 Sep 2025   Prob (F-statistic):           2.44e-07
Time:                        21:22:53   Log-Likelihood:                -30.242
No. Observations:                  32   AIC:                             82.48
Df Residuals:                      21   BIC:                             98.61
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.8488      0.136    131.375      0.0

### Modelo con train / test qsec

In [22]:
#Definir "x", "y"/ Separar en train y test

from sklearn.model_selection import train_test_split

X = df2.drop(columns = ["model", "qsec"])
y = df2['qsec']  

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4, random_state=137)

# Escalamiento de variables x

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X_train)

X_scaled_train = scaler.transform(X_train)

X_scaled_test = scaler.transform(X_test)

# Regresión

from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm

# Agregar constante para el término independiente
X_scaled_train_const = sm.add_constant(X_scaled_train)
X_scaled_test_const = sm.add_constant(X_scaled_test)

# Modelo train
modelo_train = sm.OLS(y_train, X_scaled_train_const).fit()

print(modelo_train.summary())

# Evaluación del modelo entrenado con TRAIN en TEST

y_pred_test = modelo_train.predict(X_scaled_test_const)

r2 = r2_score(y_test, y_pred_test)


print("")
print("")
print("")
print("")
print(f"R² de test: {r2:.4f}")

                            OLS Regression Results                            
Dep. Variable:                   qsec   R-squared:                       0.979
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     4.609
Date:                Fri, 12 Sep 2025   Prob (F-statistic):              0.349
Time:                        21:25:08   Log-Likelihood:               -0.25847
No. Observations:                  12   AIC:                             22.52
Df Residuals:                       1   BIC:                             27.85
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733      0.247     70.269      0.0

C:\Users\juanm\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)


### Regularizacion L2 qsec

In [25]:
from sklearn.linear_model import Ridge

# Entrenar el modelo Ridge 
modelo_ridge = Ridge(alpha=16)
modelo_ridge.fit(X_scaled_train, y_train)

# Predicciones
y_train_pred_ridge = modelo_ridge.predict(X_scaled_train)
y_test_pred_ridge = modelo_ridge.predict(X_scaled_test)

# R2
r2_train_ridge = r2_score(y_train, y_train_pred_ridge)
r2_test_ridge = r2_score(y_test, y_test_pred_ridge)

# Resultados Ridge
print("Coeficientes Ridge:", modelo_ridge.coef_)
print("Intercepto Ridge:", modelo_ridge.intercept_)
print(f"R² en entrenamiento: {r2_train_ridge:.4f}")
print(f"R² en prueba:        {r2_test_ridge:.4f}")


Coeficientes Ridge: [ 0.14304897 -0.23577204 -0.10029048 -0.29318788 -0.16568847  0.08242466
  0.27180502 -0.14769145 -0.21695304 -0.24463552]
Intercepto Ridge: 17.37333333333333
R² en entrenamiento: 0.8367
R² en prueba:        0.5698


### 2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

- Calcula el R2 e interpreta los signos de los betas.
- Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

### Modelo dummies sin train / test

In [29]:
#EDA

import pandas as pd

df3 = pd.read_excel("Motor Trend Car Road Tests.xlsx")

# print(df3.head())
# print(df3.describe())
# print(df3.info())



#Columnas dummies

df3 = pd.get_dummies(df3, columns=["cyl", "gear", "carb"], drop_first=False)

#Definir "x", "y"

from sklearn.model_selection import train_test_split

X = df3.drop(columns = ["model", "mpg"])
y = df3['mpg'] 

# Escalamiento de variables x

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X)

X_scaled = scaler.transform(X)

# Regresión

from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm

# Agregar constante para el término independiente
X_scaled_const = sm.add_constant(X_scaled)

# Modelo 21
modelo_21 = sm.OLS(y, X_scaled_const).fit()

print(modelo_21.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     7.830
Date:                Fri, 12 Sep 2025   Prob (F-statistic):           0.000124
Time:                        21:36:38   Log-Likelihood:                -66.608
No. Observations:                  32   AIC:                             167.2
Df Residuals:                      15   BIC:                             192.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         20.0906      0.501     40.114      0.0

### Modelo dummies con train / test

In [10]:
#columnas dummies
df3 = pd.read_excel("Motor Trend Car Road Tests.xlsx")
df3 = pd.get_dummies(df3, columns=["cyl", "gear", "carb"], drop_first=False)

#definir "x" "y"
X = df3.drop(columns=["model", "mpg"])
y = df3["mpg"]

#separar en train y test
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4, random_state=137)

#scalar variables
scaler = StandardScaler().fit(X_train)
X_scaled_train = scaler.transform(X_train)
X_scaled_test = scaler.transform(X_test)


#convertir array a df
X_scaled_train_df = pd.DataFrame(X_scaled_train, index=X_train.index, columns=X_train.columns)
X_scaled_test_df = pd.DataFrame(X_scaled_test, index=X_test.index, columns=X_test.columns)

#añadir constante para el modelo OLS
X_scaled_train_const = sm.add_constant(X_scaled_train_df)
X_scaled_test_const = sm.add_constant(X_scaled_test_df)

# Alinear columnas test a train 
X_scaled_test_const = X_scaled_test_const.reindex(columns=X_scaled_train_const.columns, fill_value=0)

#modelo
modelo_train = sm.OLS(y_train, X_scaled_train_const).fit()

print(modelo_train.summary())

#modelo train evaluado en test
y_pred_test = modelo_train.predict(X_scaled_test_const)

r2 = r2_score(y_test, y_pred_test)
print(f"\nR² de test: {r2:.4f}")


                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Fri, 12 Sep 2025   Prob (F-statistic):                nan
Time:                        21:58:01   Log-Likelihood:                 364.77
No. Observations:                  12   AIC:                            -705.5
Df Residuals:                       0   BIC:                            -699.7
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333        inf          0        n

C:\Users\juanm\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)
C:\Users\juanm\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1796: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\juanm\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1796: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\juanm\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1718: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
C:\Users\juanm\anaconda3\Lib\site-packages\statsmodels\base\model.py:1527: RuntimeWarning: invalid value encountered in multiply
  cov_p = self.normalized_cov_params * scale


### 2.2 Repite el ejercicio anterior usando 'qsec' como salida.

### Modelo dummies qsec

In [13]:
#EDA

import pandas as pd

df4 = pd.read_excel("Motor Trend Car Road Tests.xlsx")

# print(df3.head())
# print(df3.describe())
# print(df3.info())



#Columnas dummies

df4 = pd.get_dummies(df4, columns=["cyl", "gear", "carb"], drop_first=False)

#Definir "x", "y"

from sklearn.model_selection import train_test_split

X = df4.drop(columns = ["model", "qsec"])
y = df4['qsec'] 

# Escalamiento de variables x

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X)

X_scaled = scaler.transform(X)

# Regresión

from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm

# Agregar constante para el término independiente
X_scaled_const = sm.add_constant(X_scaled)

# Modelo 22
modelo_22 = sm.OLS(y, X_scaled_const).fit()

print(modelo_22.summary())

                            OLS Regression Results                            
Dep. Variable:                   qsec   R-squared:                       0.908
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     9.283
Date:                Fri, 12 Sep 2025   Prob (F-statistic):           4.35e-05
Time:                        22:01:24   Log-Likelihood:                -25.252
No. Observations:                  32   AIC:                             84.50
Df Residuals:                      15   BIC:                             109.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.8488      0.138    129.771      0.0

### Modelo dummies qsec train / tests

In [26]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score
import statsmodels.api as sm

#columnas dummies
df4 = pd.read_excel("Motor Trend Car Road Tests.xlsx")
df4 = pd.get_dummies(df4, columns=["cyl", "gear", "carb"], drop_first=False)

#definir "x" "y"
X = df4.drop(columns=["model", "qsec"])
y = df4["qsec"]

#separar en train y test
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4, random_state=137)

#scalar variables
scaler = StandardScaler().fit(X_train)
X_scaled_train = scaler.transform(X_train)
X_scaled_test = scaler.transform(X_test)


#convertir array a df
X_scaled_train_df = pd.DataFrame(X_scaled_train, index=X_train.index, columns=X_train.columns)
X_scaled_test_df = pd.DataFrame(X_scaled_test, index=X_test.index, columns=X_test.columns)

#añadir constante para el modelo OLS
X_scaled_train_const = sm.add_constant(X_scaled_train_df)
X_scaled_test_const = sm.add_constant(X_scaled_test_df)

# Alinear columnas test a train 
X_scaled_test_const = X_scaled_test_const.reindex(columns=X_scaled_train_const.columns, fill_value=0)

#modelo
modelo_train = sm.OLS(y_train, X_scaled_train_const).fit()

print(modelo_train.summary())

#modelo train evaluado en test
y_pred_test = modelo_train.predict(X_scaled_test_const)

r2 = r2_score(y_test, y_pred_test)
print(f"\nR² de test: {r2:.4f}")

                            OLS Regression Results                            
Dep. Variable:                   qsec   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Sat, 13 Sep 2025   Prob (F-statistic):                nan
Time:                        09:36:16   Log-Likelihood:                 363.63
No. Observations:                  12   AIC:                            -703.3
Df Residuals:                       0   BIC:                            -697.4
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733        inf          0        n

C:\Users\juanm\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)
C:\Users\juanm\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1796: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\juanm\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1796: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\juanm\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1718: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
C:\Users\juanm\anaconda3\Lib\site-packages\statsmodels\base\model.py:1527: RuntimeWarning: invalid value encountered in multiply
  cov_p = self.normalized_cov_params * scale


### 3.1 Compara los R2 de los ejercicios 1.1 & 2.1.

### 3.2 Compara los R2 de los ejercicios 1.2 & 2.2.

# Resultados y conclusiones

In [29]:
pd.read_excel("Resultados T05.xlsx")

Unnamed: 0  Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4  \
0   MODELO / VARIABLES  intercepto        mpg        cyl       disp   
1           1.1 normal     20.0906        NaN    -0.1959     1.6267   
2            1.1 train     17.4333        NaN    40.7571    -2.4635   
3             1.1 test         NaN        NaN        NaN        NaN   
4         1.1 L2 train       17.43        NaN     -0.904   -0.75512   
5          1.1 L2 test         NaN        NaN        NaN        NaN   
6                  NaN         NaN        NaN        NaN        NaN   
7           1.2 normal     17.8488     0.4096    -0.6375     -0.915   
8            1.2 train     17.3733     1.0235    -0.3081     0.2789   
9             1.2 test         NaN        NaN        NaN        NaN   
10        1.2 L2 train     17.3733      0.143    -0.2357    -0.1003   
11         1.2 L2 test         NaN        NaN        NaN        NaN   
12                 NaN         NaN        NaN        NaN        NaN   
13  2.1 dummies normal     20.0906        NaN        NaN        NaN   
14   2.1 dummies train     17.4333        NaN        NaN        NaN   
15    2.1 tdummies est         NaN        NaN        NaN        NaN   
16                 NaN         NaN        NaN        NaN        NaN   
17  2.2 dummies normal     17.8488        NaN        NaN        NaN   
18   2.2 dummies train     17.3733        NaN        NaN        NaN   
19    2.2 dummies test         NaN        NaN        NaN        NaN   

   Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9 Unnamed: 10  \
0          hp       drat         wt       qsec         vs          am   
1     -1.4497     0.4142     -3.578      1.444     0.1576      1.2378   
2    -10.1581    14.1135    -0.8005     1.5727    32.4926     36.1394   
3         NaN        NaN        NaN        NaN        NaN         NaN   
4     -0.4737     0.6741    -0.7122     0.4607     0.4402      0.6251   
5         NaN        NaN        NaN        NaN        NaN         NaN   
6         NaN        NaN        NaN        NaN        NaN         NaN   
7     -0.1054     -0.069      1.441        NaN     0.4812     -0.4426   
8    -1.01217    -1.0356     0.5976        NaN     0.2129     -0.3877   
9         NaN        NaN        NaN        NaN        NaN         NaN   
10    -0.2931    -0.1657    0.08242        NaN     0.2718     -0.1476   
11        NaN        NaN        NaN        NaN        NaN         NaN   
12        NaN        NaN        NaN        NaN        NaN         NaN   
13        NaN        NaN        NaN        NaN        NaN         NaN   
14        NaN        NaN        NaN        NaN        NaN         NaN   
15        NaN        NaN        NaN        NaN        NaN         NaN   
16        NaN        NaN        NaN        NaN        NaN         NaN   
17        NaN        NaN        NaN        NaN        NaN         NaN   
18        NaN        NaN        NaN        NaN        NaN         NaN   
19        NaN        NaN        NaN        NaN        NaN         NaN   

   Unnamed: 11 Unnamed: 12 Unnamed: 13  
0         gear        carb          R2  
1        0.476      -0.317       0.869  
2     -38.3551     11.9707        0.99  
3          NaN         NaN    -54.0719  
4       0.1347     -0.4323      0.8355  
5          NaN         NaN      0.7648  
6          NaN         NaN         NaN  
7      -0.1462      -0.435       0.875  
8       0.6755     -0.1562       0.979  
9          NaN         NaN      0.3761  
10     -0.2169     -0.2446      0.8367  
11         NaN         NaN      0.5698  
12         NaN         NaN         NaN  
13         NaN         NaN       0.893  
14         NaN         NaN           1  
15         NaN         NaN     -8.4476  
16         NaN         NaN         NaN  
17         NaN         NaN       0.908  
18         NaN         NaN           1  
19         NaN         NaN   -100.3183

### 1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

### Calcula el R2 e interpreta los signos de los betas.

El r2 de mi primer modelo me da 0.86, lo cual me dice que el modelo puede explicar o predecir en un 86% la volatilidad del mpg.

Si el signo de mi beta es positivo auemnta esa variable por lo que se consume menos gasolina y si es negativo disminuye y si consume mas gasolina por milla


### Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

En el train me da 0.99 y en el test -0.54 y segun yo el modelo esta macheteando los datos de train osea que esta overfitting y solo aprendiendo de esos datos y no el comportsmiento por lo que el desempeño de mi test es terrible

### Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

Con la penalizacion L2 me da un train de 0.83 y un test de 0.76 y creo que l darme parecido las r2 el modelo si aprendio y puede ser efectivo para estimar la varianza de los datos alrededor en un 80%

### 1.2 Repite el ejercicio anterior usando 'qsec' como salida.

Yo pienso que el modelo sufre de overfitting, ya que tiene un R2 muy alto en entrenamiento (0.97) pero bajo en test (0.37). Aunque el modelo normal da 0.83, creo que no generaliza bien y sería mejor aplicar regularización.

### 2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

### Calcula el R2 e interpreta los signos de los betas.

El modelo normal me da un R2 de 0.89 y creo que es bastante bueno paraexplicar o estimar valores.

### Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

SIn embargo cuando separo en train y test me da uun r2 de 1 y -8.44 respectivamente, investigue y esto ocurre porque cuando hago dummies me salen mas variableso betas que estimar y tcuando lo separo en 40% de train , pues lo que pasa es que hay mas incognitas que registros y por eso queda muy mal el resultado

### 2.2 Repite el ejercicio anterior usando 'qsec' como salida.

Pasa exactamente lo mismo que cuando usamos dummies para variable de salida mpg.

El modelo normal tiene un r2 decente, pero al ser train y test no se puede hacer el modelo porqeu hay mas incognitas que registros

### 3.1 Compara los R2 de los ejercicios 1.1 & 2.1.

En el primer modelo, creo que explica bastante bien el mpg, porque puede predecir bien la mayoría de la variabilidad. Si el coeficiente es positivo, eso significa que la variable hace que el consumo de gasolina baje, y si es negativo, que el consumo sube.

En el segundo modelo, al usar variables dummies, el ajuste parece mejor, pero cuando separo los datos en entrenamiento y prueba, el modelo funciona perfecto con los datos de entrenamiento pero muy mal con los de prueba. Creo que pasa porque al crear tantas variables nuevas, hay más cosas que aprender que datos para hacerlo, y eso hace que el modelo no generalice bien.

### 3.2 Compara los R2 de los ejercicios 1.2 & 2.2.

En el ejercicio 1.2, yo pienso que el modelo está sobreajustado porque funciona muy bien con los datos de entrenamiento pero mal con los de prueba. Aunque el modelo normal tiene un resultado decente, no generaliza bien y creo que se debería usar regularización.

En el ejercicio 2.2 pasa lo mismo que con las dummies para mpg: el modelo parece bueno en general, pero cuando se divide en entrenamiento y prueba no funciona bien porque hay más variables que datos, y eso hace que el modelo no pueda aprender correctamente.